src/main/java/org/jenkinsci/plugins/workflow/graphanalysis/AbstractFlowScanner.java

```diff
@@ -0,0 +1,255 @@
+    protected abstract void setHeads(@Nonnull Collection<FlowNode> filteredHeads);
+
+    /**
+     * Actual meat of the iteration, get the next node to visit, using & updating state as needed
+     * @param current Current node to use in generating next value
+     * @param blackList Nodes that are not eligible for visiting
+     * @return Next node to visit, or null if we've exhausted the node list
+     */
+    @CheckForNull
+    protected abstract FlowNode next(@Nonnull FlowNode current, @Nonnull Collection<FlowNode> blackList);
+
+    @Override
+    public boolean hasNext() {
+        return _next != null;
+    }
+
+    @Override
+    public FlowNode next() {
+        if (_next == null) {
+            throw new NoSuchElementException();
+        }
+
+        // For computing timings and changes, it may be helpful to keep the previous result
+        // by creating a variable _last and storing _current to it.
+
+//            System.out.println("Current iterator val: " + ((_current == null) ? "null" : _current.getId()));
+//            System.out.println("Next iterator val: " + ((_next == null) ? "null" : _next.getId()));
+        _current = _next;
+        _next = next(_current, _blackList);
+//            System.out.println("New next val: " + ((_next == null) ? "null" : _next.getId()));
+        return _current;
+    }
+
+    @Override
+    public void remove() {
+        throw new UnsupportedOperationException("FlowGraphs are immutable, so FlowScanners can't remove nodes");
+    }
+
+    @Override
+    public Iterator<FlowNode> iterator() {
+        return this;
+    }
+
+    public Filterator<FlowNode> filter(Predicate<FlowNode> filterCondition) {
+        return new FilteratorImpl<FlowNode>(this, filterCondition);
+    }
+
+    // Basic algo impl
+    public FlowNode findFirstMatch(@CheckForNull Collection<FlowNode> heads,
+                                           @CheckForNull Collection<FlowNode> endNodes,
+                                           Predicate<FlowNode> matchCondition) {
+        if (!setup(heads, endNodes)) {
+            return null;
+        }
+
+        for (FlowNode f : this) {
+            if (matchCondition.apply(f)) {
+                return f;
+            }
+        }
+        return null;
+    }
+
+    // Polymorphic methods for syntactic sugar
+
+    @CheckForNull
+    public FlowNode findFirstMatch(@CheckForNull Collection<FlowNode> heads, @Nonnull Predicate<FlowNode> matchPredicate) {
+        return this.findFirstMatch(heads, null, matchPredicate);
+    }
+
+    @CheckForNull
+    public FlowNode findFirstMatch(@CheckForNull FlowNode head, @Nonnull Predicate<FlowNode> matchPredicate) {
+        return this.findFirstMatch(Collections.singleton(head), null, matchPredicate);
+    }
+
+    @CheckForNull
+    public FlowNode findFirstMatch(@CheckForNull FlowExecution exec, @Nonnull Predicate<FlowNode> matchPredicate) {
+        if (exec != null && exec.getCurrentHeads() != null) {
```

Comment:
:ant: do not allow `exec` to be null, and :bug: `getCurrentHeads` will never be null.